File: amplify/agent/runtime/exports/deck_exporter.py
"""折衷デッキのエクスポート（HTML組み立て・PNG/PDF/PPTX/サムネイル生成）"""

import html as html_escape
import io
import json
import re
import subprocess
import tempfile
from pathlib import Path

_ASSET_DIR = Path(__file__).parent.parent / "decks" / "eclectic"
_RENDER_SCRIPT = Path(__file__).parent / "render_deck.mjs"

RENDER_TIMEOUT = 300  # Chromiumレンダリングのタイムアウト（秒）


def is_deck_source(source: str) -> bool:
    """スライドソースが折衷デッキ（HTML）かどうかを判定"""
    stripped = source.lstrip()
    # HTMLコメントをスキップして先頭要素を判定
    stripped = re.sub(r'^(<!--.*?-->\s*)+', '', stripped, flags=re.DOTALL)
    return stripped.startswith('<')


def extract_deck_title(slides_html: str) -> str | None:
    """デッキHTMLからタイトル（最初のh1/h2のテキスト）を抽出"""
    match = re.search(r'<h[12][^>]*>(.*?)</h[12]>', slides_html, re.DOTALL)
    if not match:
        return None
    text = re.sub(r'<br\s*/?>', ' ', match.group(1))
    text = re.sub(r'<[^>]+>', '', text)
    text = html_escape.unescape(text).strip()
    return text or None


def assemble_deck_html(slides_html: str, title: str | None = None, subtitle: str = "") -> str:
    """スライドブロックをシェルに埋め込み、CSS/JSをインライン化した単一HTMLを返す"""
    shell = (_ASSET_DIR / "deck-shell.html").read_text(encoding="utf-8")
    css = (_ASSET_DIR / "deck.css").read_text(encoding="utf-8")
    js = (_ASSET_DIR / "deck.js").read_text(encoding="utf-8")

    resolved_title = title or extract_deck_title(slides_html) or "スライド"
    html = shell.replace("{{DECK_TITLE}}", html_escape.escape(resolved_title))
    html = html.replace("{{DECK_SUBTITLE}}", html_escape.escape(subtitle))

    # 外部参照をインライン化して自己完結型の1ファイルにする（共有・レンダリング両用）
    html = html.replace('<link rel="stylesheet" href="deck.css">', f"<style>\n{css}\n</style>")
    html = html.replace('<script src="deck.js"></script>', f"<script>\n{js}\n</script>")

    start_marker = "<!-- SLIDES:START -->"
    end_marker = "<!-- SLIDES:END -->"
    start = html.index(start_marker) + len(start_marker)
    end = html.index(end_marker)
    return html[:start] + "\n" + slides_html + "\n" + html[end:]


def _render_deck_pngs(slides_html: str, scale: int = 2, limit: int | None = None) -> tuple[list[Path], list[dict]]:
    """デッキをChromiumでレンダリングし、PNGパスのリストとノート情報を返す"""
    tmpdir = Path(tempfile.mkdtemp())
    deck_path = tmpdir / "deck.html"
    deck_path.write_text(assemble_deck_html(slides_html), encoding="utf-8")
    out_dir = tmpdir / "frames"

    cmd = ["node", str(_RENDER_SCRIPT), str(deck_path), str(out_dir), "--scale", str(scale)]
    if limit is not None:
        cmd.extend(["--limit", str(limit)])

    result = subprocess.run(cmd, capture_output=True, text=True, timeout=RENDER_TIMEOUT)
    if result.returncode != 0:
        raise RuntimeError(f"Deck render error: {result.stderr or result.stdout}")

    pngs = sorted(out_dir.glob("slide-*.png"))
    if not pngs:
        raise RuntimeError("Deck render failed: no PNG files created")

    notes_path = out_dir / "notes.json"
    notes = json.loads(notes_path.read_text(encoding="utf-8")) if notes_path.exists() else []
    return pngs, notes


def generate_deck_pptx(slides_html: str, title: str | None = None) -> bytes:
    """折衷デッキをPPTXに変換（全面画像スライド＋スピーカーノート）"""
    from pptx import Presentation
    from pptx.util import Inches

    pngs, notes = _render_deck_pngs(slides_html)
    notes_by_index = {int(item["index"]): item for item in notes}

    prs = Presentation()
    # 1920×1080のデザインに合わせた16:9キャンバス
    prs.slide_width = Inches(13.333)
    prs.slide_height = Inches(7.5)
    blank = prs.slide_layouts[6]

    for i, png in enumerate(pngs, start=1):
        slide = prs.slides.add_slide(blank)
        slide.shapes.add_picture(str(png), 0, 0, width=prs.slide_width, height=prs.slide_height)
        meta = notes_by_index.get(i)
        if meta and meta.get("notes"):
            tf = slide.notes_slide.notes_text_frame
            label = meta.get("label", "")
            tf.text = (f"[{label}] " if label else "") + meta["notes"]

    prs.core_properties.title = title or extract_deck_title(slides_html) or "スライド"

    buf = io.BytesIO()
    prs.save(buf)
    return buf.getvalue()


def generate_deck_pdf(slides_html: str) -> bytes:
    """折衷デッキをPDFに変換（レンダリング済みPNGを結合）"""
    import img2pdf

    pngs, _ = _render_deck_pngs(slides_html)
    return img2pdf.convert([str(p) for p in pngs])


def generate_deck_thumbnail(slides_html: str) -> bytes:
    """折衷デッキの1枚目をPNGで生成（OGP用サムネイル）"""
    pngs, _ = _render_deck_pngs(slides_html, scale=1, limit=1)
    return pngs[0].read_bytes()


def generate_deck_standalone_html(slides_html: str, title: str | None = None) -> str:
    """折衷デッキのスタンドアロンHTMLを生成（共有用）"""
    return assemble_deck_html(slides_html, title=title)

File: amplify/agent/runtime/src/exports/deckExporter.ts
/** 折衷デッキのエクスポート（HTML組み立て・PNG/PDF/PPTX/サムネイル生成）
 *
 * Python版 deck_exporter.py + render_deck.mjs の移植。
 * ランタイムがNodeになったため、Chromiumレンダリングはサブプロセスではなく
 * playwright-core をインプロセスで使用する。
 */

import { readFileSync, existsSync } from 'node:fs';
import { mkdtemp, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { fileURLToPath, pathToFileURL } from 'node:url';
import path from 'node:path';
import { chromium } from 'playwright-core';
import PptxGenJSImport from 'pptxgenjs';

// pptxgenjs はCJSパッケージのため、NodeNext環境では実行時のdefault二重ラップを解き、
// 型はこのファイルで使う操作だけを定義する
interface PptxSlide {
  addImage(options: { data: string; x: number; y: number; w: number; h: number }): void;
  addNotes(notes: string): void;
}
interface PptxGen {
  defineLayout(options: { name: string; width: number; height: number }): void;
  layout: string;
  title: string;
  addSlide(): PptxSlide;
  write(options: { outputType: string }): Promise<unknown>;
}
const PptxGenJS = (
  (PptxGenJSImport as unknown as { default?: unknown }).default ?? PptxGenJSImport
) as new () => PptxGen;
import { PDFDocument } from 'pdf-lib';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// dist/exports からも src/exports からも runtime/decks を指す
const ASSET_DIR = path.resolve(__dirname, '..', '..', 'decks', 'eclectic');

const DESIGN_WIDTH = 1920;
const DESIGN_HEIGHT = 1080;

function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

/** スライドソースが折衷デッキ（HTML）かどうかを判定 */
export function isDeckSource(source: string): boolean {
  // 先頭のHTMLコメントをスキップして先頭要素を判定
  const stripped = source.trimStart().replace(/^(<!--[\s\S]*?-->\s*)+/, '');
  return stripped.startsWith('<');
}

/** デッキHTMLからタイトル（最初のh1/h2のテキスト）を抽出 */
export function extractDeckTitle(slidesHtml: string): string | null {
  const match = slidesHtml.match(/<h[12][^>]*>([\s\S]*?)<\/h[12]>/);
  if (!match) return null;
  let text = match[1].replace(/<br\s*\/?>/g, ' ');
  text = text.replace(/<[^>]+>/g, '');
  text = text
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .trim();
  return text || null;
}

/** スライドブロックをシェルに埋め込み、CSS/JSをインライン化した単一HTMLを返す */
export function assembleDeckHtml(slidesHtml: string, title?: string | null, subtitle: string = ''): string {
  const shell = readFileSync(path.join(ASSET_DIR, 'deck-shell.html'), 'utf-8');
  const css = readFileSync(path.join(ASSET_DIR, 'deck.css'), 'utf-8');
  const js = readFileSync(path.join(ASSET_DIR, 'deck.js'), 'utf-8');

  const resolvedTitle = title || extractDeckTitle(slidesHtml) || 'スライド';
  let html = shell.replace('{{DECK_TITLE}}', escapeHtml(resolvedTitle));
  html = html.replace('{{DECK_SUBTITLE}}', escapeHtml(subtitle));

  // 外部参照をインライン化して自己完結型の1ファイルにする（共有・レンダリング両用）
  html = html.replace('<link rel="stylesheet" href="deck.css">', `<style>\n${css}\n</style>`);
  html = html.replace('<script src="deck.js"></script>', `<script>\n${js}\n</script>`);

  const startMarker = '<!-- SLIDES:START -->';
  const endMarker = '<!-- SLIDES:END -->';
  const start = html.indexOf(startMarker) + startMarker.length;
  const end = html.indexOf(endMarker);
  return html.slice(0, start) + '\n' + slidesHtml + '\n' + html.slice(end);
}

function findChromium(): string {
  const candidates = [
    process.env.PUPPETEER_EXECUTABLE_PATH,
    '/usr/bin/chromium',
    '/usr/bin/chromium-browser',
    '/usr/bin/google-chrome',
  ];
  for (const c of candidates) {
    if (c && existsSync(c)) return c;
  }
  throw new Error('Chromium not found (set PUPPETEER_EXECUTABLE_PATH).');
}

export interface SlideNote {
  index: number;
  label: string;
  notes: string;
}

interface RenderResult {
  pngs: Buffer[];
  notes: SlideNote[];
}

/** デッキをChromiumでレンダリングし、各スライドのPNGとノート情報を返す */
async function renderDeckPngs(slidesHtml: string, scale: number = 2, limit?: number): Promise<RenderResult> {
  const dir = await mkdtemp(path.join(tmpdir(), 'deck-'));
  const deckPath = path.join(dir, 'deck.html');
  await writeFile(deckPath, assembleDeckHtml(slidesHtml), 'utf-8');

  // AgentCoreコンテナはroot実行のためsandboxを無効化
  const browser = await chromium.launch({
    executablePath: findChromium(),
    args: ['--no-sandbox', '--disable-dev-shm-usage', '--disable-gpu'],
  });

  try {
    const page = await browser.newPage({
      viewport: { width: DESIGN_WIDTH, height: DESIGN_HEIGHT },
      deviceScaleFactor: scale,
    });
    await page
      .goto(pathToFileURL(deckPath).href, { waitUntil: 'networkidle' })
      .catch((e) => console.warn(`[WARN] Deck navigation issue: ${e.message}`));
    // Webフォント（Shippori Mincho等）の読み込みを待つ。オフライン時はフォールバックで続行
    await page.waitForTimeout(600);
    await page
      .evaluate(() => (document as unknown as { fonts?: { ready: Promise<unknown> } }).fonts?.ready)
      .catch((e) => console.warn(`[WARN] Font wait failed: ${e}`));

    // キャプチャ用にフラット化: フィットスケールを無効化し、プレゼン用UIを隠し、
    // スライドを1枚ずつ表示する
    await page.addStyleTag({
      content: `
      #stage-scaler{position:absolute!important;left:0!important;top:0!important;
        transform:none!important;width:${DESIGN_WIDTH}px!important;height:${DESIGN_HEIGHT}px!important;}
      .deck-slide{position:absolute!important;inset:0!important;display:none!important;}
      .deck-slide.__cap{display:block!important;}
      .progress,.hud,.notes,.overview,.help{display:none!important;}
      html,body{background:#F5F2EC!important;overflow:hidden!important;}
    `,
    });

    const slideCount = await page.locator('.deck-slide').count();
    if (slideCount === 0) {
      throw new Error('No .deck-slide elements found — the deck failed to load.');
    }
    const total = limit !== undefined ? Math.min(slideCount, limit) : slideCount;

    const pngs: Buffer[] = [];
    const notes: SlideNote[] = [];

    for (let i = 0; i < total; i++) {
      const info = await page.evaluate((idx) => {
        const all = document.querySelectorAll('.deck-slide');
        all.forEach((el, k) => el.classList.toggle('__cap', k === idx));
        const sec = all[idx].querySelector('section') as HTMLElement | null;
        return {
          label: sec?.dataset.label ?? '',
          notes: sec?.dataset.speakerNotes ?? '',
        };
      }, i);
      await page.waitForTimeout(120);
      const png = await page.screenshot({
        clip: { x: 0, y: 0, width: DESIGN_WIDTH, height: DESIGN_HEIGHT },
      });
      pngs.push(png);
      notes.push({ index: i + 1, label: info.label, notes: info.notes });
    }

    return { pngs, notes };
  } finally {
    await browser.close();
  }
}

/** 折衷デッキをPPTXに変換（全面画像スライド＋スピーカーノート） */
export async function generateDeckPptx(slidesHtml: string, title?: string | null): Promise<Buffer> {
  const { pngs, notes } = await renderDeckPngs(slidesHtml);

  const pptx = new PptxGenJS();
  // 1920×1080のデザインに合わせた16:9キャンバス（13.333in × 7.5in）
  pptx.defineLayout({ name: 'DECK_16x9', width: 13.333, height: 7.5 });
  pptx.layout = 'DECK_16x9';
  pptx.title = title || extractDeckTitle(slidesHtml) || 'スライド';

  pngs.forEach((png, i) => {
    const slide = pptx.addSlide();
    slide.addImage({
      data: `image/png;base64,${png.toString('base64')}`,
      x: 0,
      y: 0,
      w: 13.333,
      h: 7.5,
    });
    const meta = notes[i];
    if (meta?.notes) {
      slide.addNotes((meta.label ? `[${meta.label}] ` : '') + meta.notes);
    }
  });

  return (await pptx.write({ outputType: 'nodebuffer' })) as Buffer;
}

/** 折衷デッキをPDFに変換（レンダリング済みPNGを結合） */
export async function generateDeckPdf(slidesHtml: string): Promise<Buffer> {
  const { pngs } = await renderDeckPngs(slidesHtml);

  const doc = await PDFDocument.create();
  for (const png of pngs) {
    const image = await doc.embedPng(png);
    const page = doc.addPage([DESIGN_WIDTH, DESIGN_HEIGHT]);
    page.drawImage(image, { x: 0, y: 0, width: DESIGN_WIDTH, height: DESIGN_HEIGHT });
  }
  return Buffer.from(await doc.save());
}

/** 折衷デッキの1枚目をPNGで生成（OGP用サムネイル） */
export async function generateDeckThumbnail(slidesHtml: string): Promise<Buffer> {
  const { pngs } = await renderDeckPngs(slidesHtml, 1, 1);
  return pngs[0];
}

/** 折衷デッキのスタンドアロンHTMLを生成（共有用） */
export function generateDeckStandaloneHtml(slidesHtml: string, title?: string | null): string {
  return assembleDeckHtml(slidesHtml, title);
}
